lib/grammers-client/src/client/auth.rs

```diff
@@ -14,25 +14,51 @@ use grammers_tl_types as tl;
 use std::convert::TryInto;
 use std::fmt;
 
+use grammers_crypto::two_factor_auth::{calculate_2fa, VerificationError};
+
 /// The error type which is returned when signing in fails.
 #[derive(Debug)]
 pub enum SignInError {
-    SignUpRequired {
-        terms_of_service: Option<tl::types::help::TermsOfService>,
-    },
     InvalidCode,
+    IncorrectParameters(VerificationError),
+    IncorrectPassword,
     Other(InvocationError),
 }
 
 impl fmt::Display for SignInError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use SignInError::*;
         match self {
+            InvalidCode => write!(f, "sign in error: invalid code"),
+            IncorrectParameters(_e) => write!(f, "Incorrect parameters for password verification received by telegram. Please request them again."),
+            IncorrectPassword => write!(f, "Incorrect password"),
+            Other(e) => write!(f, "sign in error: {}", e),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum SignInResult {
+    SignedIn(tl::types::User),
+    PasswordRequired {
+        password_information: tl::types::account::Password,
+    },
```

Comment:
I think I would prefer this to be something like the following:

```suggestion
    PasswordRequired(PasswordToken),
```

This is more in line with how `send_code_request` works (it returns an opaque "token" you can later use), and the hint can be trivially accessible with:

```rust
impl PasswordToken {
    fn hint(&self) -> Option<&str> {
        self.password.hint.as_ref()
    }
}
```

The password token simply contains the password as a private field (we want to avoid exposing TL types where possible), even if not done yet for the TOS.